traffic_ops/traffic_ops_golang/role/roles.go

```diff
@@ -273,3 +291,522 @@ RETURNING id, last_updated`
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("creating role: "+err.Error()))
+		return
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var throwaway interface{}
+		if err := rows.Scan(&roleID, &throwaway); err != nil {
+			api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, errors.New("role create: scanning role ID: "+err.Error()))
+			return
+		}
+	}
+
+	if roleCapabilities != nil && len(*roleCapabilities) > 0 {
+		userErr, sysErr, errCode = createRoleCapabilityAssociations(inf.Tx, roleID, roleCapabilities)
+		if userErr != nil || sysErr != nil {
+			api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+			return
+		}
+	}
+	alerts := tc.CreateAlerts(tc.SuccessLevel, "role was created.")
+	var roleResponse interface{}
+	if version.Major >= 5 {
+		var capabilities []string
+		if roleCapabilities != nil {
+			capabilities = *roleCapabilities
+		}
+		roleResponse = tc.RoleV50{
+			Name:        util.StrPtr(roleName),
+			Permissions: capabilities,
+			Description: util.StrPtr(roleDesc),
+		}
+	} else {
+		roleResponse = tc.Role{
+			RoleV11: tc.RoleV11{
+				ID:          util.IntPtr(roleID),
+				Name:        util.StrPtr(roleName),
+				Description: util.StrPtr(roleDesc),
+				PrivLevel:   util.IntPtr(privLevel),
+			},
+			Capabilities: roleCapabilities,
+		}
+	}
+	api.WriteAlertsObj(w, r, http.StatusOK, alerts, roleResponse)
+}
+
+func Get(w http.ResponseWriter, r *http.Request) {
+	var maxTime time.Time
+	var runSecond bool
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	tx := inf.Tx.Tx
+	version := inf.Version
+	if version == nil {
+		middleware.NotImplementedHandler().ServeHTTP(w, r)
+		return
+	}
+
+	params := make(map[string]dbhelpers.WhereColumnInfo, 0)
+	if version.Major >= 5 {
+		params["name"] = dbhelpers.WhereColumnInfo{Column: "name"}
+	} else {
+		params["name"] = dbhelpers.WhereColumnInfo{Column: "name"}
+		params["id"] = dbhelpers.WhereColumnInfo{Column: "id", Checker: api.IsInt}
+		params["privLevel"] = dbhelpers.WhereColumnInfo{Column: "priv_level", Checker: api.IsInt}
+	}
+
+	if _, ok := inf.Params["orderby"]; !ok {
+		inf.Params["orderby"] = "name"
+	}
+	where, orderBy, pagination, queryValues, errs := dbhelpers.BuildWhereAndOrderByAndPagination(inf.Params, params)
+	if len(errs) != 0 {
+		api.HandleErr(w, r, tx, http.StatusBadRequest, util.JoinErrs(errs), nil)
+		return
+	}
+	if version.Major >= 5 {
+		if perm, ok := inf.Params["can"]; ok {
+			queryValues["can"] = perm
+			where = dbhelpers.AppendWhere(where, "permissions @> :can")
+		}
+	}
+	if inf.Config.UseIMS {
+		runSecond, maxTime = ims.TryIfModifiedSinceQuery(inf.Tx, r.Header, queryValues, selectMaxLastUpdatedQuery(where))
+		if !runSecond {
+			log.Debugln("IMS HIT")
+			api.AddLastModifiedHdr(w, maxTime)
+			w.WriteHeader(http.StatusNotModified)
+			api.WriteResp(w, r, []interface{}{})
```

Comment:
Is this `api.WriteResp` call actually necessary? I see other IMS hits doing it, but idk why, since the whole point is that clients will just ignore it and use their cache.